### ISSUE TITLE
[FR] How to log metric of array

### ISSUE THREAD

**Issue by cometta** [mlflow/mlflow#2451]
how to log output from sklearn.roc_curve with mlflow.log_metric() ? The reason is the output is in array instead of single value. I would like to use the charting feature of mlflow.

**Comment by nishnash54** [comment]
Currently mlflow does not allow you to assign a metric to the `x-axis`, so it isn't possible to create a roc curve. Also as you have mentioned, it's not possible to log the entire array at once. It is a great feature request. 

For now, you can use a work around and log each element of the list by looping thought them and assigning a step. This way, you can still plot the metric on the chart but both the metrics will be along the `y-axis`.

```py
fpr, tpr, thresh = roc_curve(y, scores, pos_label=2)
with mlflow.start_run():
    for idx in range(len(fpr)):
        mlflow.log_metrics({'fpr': fpr[idx], 'tpr': tpr[idx]}, step=idx+1)
```

**Comment by smurching** [comment]
@cometta we'd be happy to consider an enhancement to the `mlflow.log_metric` API to allow logging an array of values for the metric. There are some tricky considerations though - for example, what should the `step`  & `timestamp` values of the metric be? 

One strawman is that when calling `mlflow.log_metric(key, array_of_values)`, we'd use the current timestamp as the timestamp for all metric values, and assign each metric value a step from 0 to `len(array_of_values)`.

However by default, we display metric values in the UI using the timestamp as the X-axis, so this would cause the curve to appear as a single line (could be solved in the UI by adding special-case logic to use step as the default x axis if all timestamps are the same). We'd also run into issues if people called `mlflow.log_metric(key, array_of_values)` multiple times with the same key (we'd have multiple values for the same step).

Would be happy to review a proposal that tackles some of these issues (I'd start by investigating what other use cases there are for an `mlflow.log_metric(key, array_of_values)` API) if you've got the time to investigate, thanks!

**Comment by pakelley** [comment]
I definitely think having support for an ROC curve (or PR curve) would be useful! In addition to logging an array of metric values, it would be helpful to add support for a metric on the x-axis of the metric plot view. Some thought would need to go into the UX, of course, but I think this would be useful for visualizing the ROC curve.

The values @smurching proposed for `step`/`timestamp` are what I would personally expect as defaults, and I figure different values could be optionally passed in.

As far as use cases for an array of metric values, varying threshold values seems like the obvious one. Maybe this would also be useful for making predictions with multiple models for comparison? Maybe there's a more appropriate way of doing this though.

**Comment by fulsi** [comment]
Hello, 

In the proposed workaround, if the number of fpr and tpr change between different trials I will not be possible to compare them properly (not aligned)

I propose the following workaround. As the thresold, vary between 0 and 1, it will always be possible to compare different trial.

```python
fpr, tpr, thresh = precision_recall_curve(y, scores, pos_label=2)
list_thresh = np.append(list_thresh, (1, ))  # view precision_recall_curve from sklearn metrics
with mlflow.start_run():
    for idx in range(len(fpr)):
        seuil = int(round(list_thresh[idx]*10000, 0))  # use high number to avoid several points on the same step
        mlflow.log_metrics({'precision': precision[idx], 'rappel': recall[idx]}, step=seuil)
```


Actually this workaround works for anything else, as long as the axis x keep the same interval through trials. The only constrainst is to put an integer instead of the step. It would be great to have an additionnal custom field (step / timestamp) to store whatever we want using float.

**Comment by dbczumar** [comment]
We do not have plans to extend metric logging to arrays at this time, but the example workaround provided by @fulsi is fantastic! I'm going to go ahead and close this issue for now. Thank you for using MLflow!

**Comment by milannnnn** [comment]
Are there any news/timelines for the implementation of "array" metrics? While the proposed approach (step) can be used as a workaround in some cases, it is not a perfect replacement (for example if we want to log an array of metrics at every epoch).
